server/server.go

```diff
@@ -433,3 +435,50 @@ func ServeOnLocalTailscaled(ctx context.Context, lc *local.Client, st *ipnstate.
 
 	return func() { watcher.Close() }, watcherChan, nil
 }
+
+// writeHTTPError writes a JSON error response
+func writeHTTPError(w http.ResponseWriter, r *http.Request, statusCode int, errorCode, errorDescription string) {
+	args := []any{
+		slog.Int("status", statusCode),
+		slog.String("method", r.Method),
+		slog.String("path", r.URL.Path),
+		slog.String("error", errorCode),
+		slog.String("description", errorDescription),
+	}
+
+	switch statusCode {
+	case http.StatusForbidden, http.StatusUnauthorized:
+		slog.Warn("HTTP error", args...)
+	case http.StatusInternalServerError:
+		slog.Error("HTTP error", args...)
+	default:
+		slog.Debug("HTTP error", args...)
+	}
+
+	w.Header().Set("Cache-Control", "no-store")
+	w.Header().Set("Pragma", "no-cache")
+	w.WriteHeader(statusCode)
+
+	acceptHeader := r.Header.Get("Accept")
+	switch {
+	case strings.Contains(acceptHeader, "text/html"):
+		fallthrough
+	case strings.Contains(acceptHeader, "text/plain"):
+		w.Header().Set("Content-Type", "text/plain; charset=UTF-8")
+		fmt.Fprintf(w, "Error %d: %s - %s", statusCode, errorCode, errorDescription)
+	default:
+		w.Header().Set("Content-Type", "application/json;charset=UTF-8")
+		json.NewEncoder(w).Encode(oauthErrorResponse{
```

Comment:
why are all errors encoded as `oauthErrorResponse`? should we be doing it for all errors on all handlers?

Reply:
legacy naming. I'll change it to something more abstract.